File: src/screens/RanksScreen.js
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { View, Text, StyleSheet, ScrollView, TouchableOpacity, Animated, Easing } from 'react-native';
import { useFocusEffect } from '@react-navigation/native';
import ViewShot from 'react-native-view-shot';
import { getWorkouts } from '../storage';
import { ranks, totalPoints, rankFor, levelOf, maxLevel } from '../data/ranks';
import { useTheme } from '../ThemeContext';
import { useDiscipline } from '../DisciplineContext';
import DisciplineSwitch from '../components/DisciplineSwitch';
import Medal from '../components/Medal';
import ShareCard from '../components/ShareCard';
import FadeInView from '../components/FadeInView';
import AnimatedNumber from '../components/AnimatedNumber';
import PressableScale from '../components/PressableScale';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { shareProgressImage } from '../share';
import { spacing, radius } from '../theme';

export default function RanksScreen() {
  const { theme } = useTheme();
  const { discipline } = useDiscipline();
  const styles = useMemo(() => makeStyles(theme.colors), [theme]);
  const [points, setPoints] = useState(0);
  const [count, setCount] = useState(0);
  const shotRef = useRef(null);

  useFocusEffect(
    useCallback(() => {
      let active = true;
      getWorkouts(discipline).then((w) => {
        if (!active) return;
        setPoints(totalPoints(w));
        setCount(w.length);
      });
      return () => {
        active = false;
      };
    }, [discipline])
  );

  const { current, next, progress } = rankFor(points);
  const level = levelOf(current.id);

  const onShare = () => {
    const label = discipline === 'gym' ? 'Palestra' : 'Calisthenics';
    const text = `💪 CaliStrack (${label})\nRango ${current.name} ${current.icon} — Livello ${level}/${maxLevel}\n${points} punti • ${count} allenamenti${next ? `\nProssimo: ${next.name} ${next.icon}` : ' • rango MAX!'}`;
    shareProgressImage(shotRef, text);
  };

  return (
    <View style={styles.container}>
      <DisciplineSwitch />
      <ScrollView contentContainerStyle={styles.content}>
        <FadeInView style={[styles.hero, { borderColor: current.color }]}>
          <Medal rank={current} progress={progress} size={140} />
          <Text style={[styles.heroName, { color: current.color }]}>
            {current.name}
          </Text>
          <Text style={styles.heroLevel}>
            Livello {level} / {maxLevel}
          </Text>
          <View style={styles.heroPtsRow}>
            <AnimatedNumber value={points} style={styles.heroPtsNum} />
            <Text style={styles.heroPts}> punti • {count} allenamenti</Text>
          </View>
          <Text style={styles.heroNext}>
            {next
              ? `Ancora ${next.min - points} punti per il ${next.name}`
              : 'Hai raggiunto il rango massimo!'}
          </Text>

          <PressableScale style={styles.shareBtn} onPress={onShare}>
            <MaterialCommunityIcons name="share-variant" size={18} color={theme.colors.bg} />
            <Text style={styles.shareText}>Condividi i progressi</Text>
          </PressableScale>
        </FadeInView>

        <Text style={styles.section}>Scala dei ranghi</Text>
        {[...ranks].reverse().map((r, idx) => {
          const unlocked = points >= r.min;
          const isCurrent = r.id === current.id;
          return (
            <FadeInView
              key={r.id}
              delay={60 + idx * 45}
              style={[
                styles.rankRow,
                isCurrent && { borderColor: r.color, borderWidth: 2 },
                !unlocked && styles.locked,
              ]}
            >
              <Medal
                rank={r}
                progress={unlocked ? 1 : 0}
                size={54}
                showLevel={false}
                locked={!unlocked}
              />
              <View style={{ flex: 1 }}>
                <Text
                  style={[
                    styles.rankName,
                    { color: unlocked ? r.color : theme.colors.textMuted },
                  ]}
                >
                  Lv {levelOf(r.id)} · {r.name}
                </Text>
                <Text style={styles.rankReq}>{r.min} punti</Text>
              </View>
              {isCurrent && <Text style={styles.badge}>ATTUALE</Text>}
            </FadeInView>
          );
        })}

        <Text style={styles.info}>
          Guadagni punti a ogni allenamento. Nel calisthenics: 1 punto per ripetizione
          e 0,5 per secondo di hold. In palestra: volume (kg × reps) diviso 10.
        </Text>
      </ScrollView>

      {/* Card fuori schermo, catturata per la condivisione */}
      <View style={styles.offscreen} pointerEvents="none">
        <ViewShot ref={shotRef} options={{ format: 'png', quality: 1 }}>
          <ShareCard
            rank={current}
            progress={progress}
            points={points}
            next={next}
            discipline={discipline}
            workouts={count}
          />
        </ViewShot>
      </View>
    </View>
  );
}

const makeStyles = (c) =>
  StyleSheet.create({
    container: { flex: 1, backgroundColor: c.bg },
    content: { padding: spacing.md },
    hero: {
      backgroundColor: c.card,
      borderRadius: radius.lg,
      padding: spacing.lg,
      alignItems: 'center',
      borderWidth: 2,
    },
    heroName: { fontSize: 26, fontWeight: '800', marginTop: spacing.md },
    heroLevel: { color: c.text, fontSize: 15, fontWeight: '700', marginTop: 2 },
    heroPtsRow: { flexDirection: 'row', alignItems: 'baseline', marginTop: spacing.xs },
    heroPtsNum: { color: c.primary, fontWeight: '900', fontSize: 16 },
    heroPts: { color: c.textMuted },
    heroNext: { color: c.textMuted, marginTop: spacing.sm, textAlign: 'center' },
    shareBtn: {
      marginTop: spacing.md,
      backgroundColor: c.primary,
      borderRadius: radius.md,
      paddingVertical: spacing.sm,
      paddingHorizontal: spacing.lg,
      flexDirection: 'row',
      alignItems: 'center',
      gap: spacing.sm,
    },
    shareText: { color: c.bg, fontWeight: '800' },
    section: {
      color: c.text,
      fontSize: 16,
      fontWeight: '700',
      marginTop: spacing.lg,
      marginBottom: spacing.sm,
    },
    rankRow: {
      flexDirection: 'row',
      alignItems: 'center',
      gap: spacing.md,
      backgroundColor: c.card,
      borderRadius: radius.md,
      padding: spacing.md,
      marginBottom: spacing.sm,
      borderWidth: 1,
      borderColor: c.border,
    },
    locked: { opacity: 0.55 },
    rankName: { fontSize: 16, fontWeight: '700' },
    rankReq: { color: c.textMuted, fontSize: 12, marginTop: 2 },
    badge: {
      color: c.bg,
      backgroundColor: c.primary,
      fontSize: 10,
      fontWeight: '800',
      paddingHorizontal: spacing.sm,
      paddingVertical: 3,
      borderRadius: radius.sm,
      overflow: 'hidden',
    },
    info: { color: c.textMuted, fontSize: 12, marginTop: spacing.md, lineHeight: 18 },
    offscreen: { position: 'absolute', left: -1000, top: 0 },
  });

File: lib/screens/ranks_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:share_plus/share_plus.dart';
import '../state/theme_provider.dart';
import '../state/discipline_provider.dart';
import '../state/workout_provider.dart';
import '../state/locale_provider.dart';
import '../services/feedback_service.dart';
import '../data/ranks.dart';
import '../theme/app_theme.dart';
import '../widgets/discipline_switch.dart';
import '../widgets/medal.dart';

class RanksScreen extends StatelessWidget {
  const RanksScreen({super.key});

  @override
  Widget build(BuildContext context) {
    final c = context.watch<ThemeProvider>().colors;
    final t = context.watch<LocaleProvider>().t;
    final discipline = context.watch<DisciplineProvider>().discipline;
    final workouts = context.watch<WorkoutProvider>().forDiscipline(discipline);
    final pts = totalPoints(workouts);
    final info = rankFor(pts);
    final level = levelOf(info.current.id);

    return Scaffold(
      appBar: AppBar(title: Text(t('nav_medals'))),
      body: ListView(
        padding: const EdgeInsets.all(Spacing.md),
        children: [
          const DisciplineSwitch(),
          const SizedBox(height: Spacing.sm),
          // Hero card
          Container(
            padding: const EdgeInsets.all(Spacing.lg),
            decoration: BoxDecoration(
              color: c.card,
              borderRadius: BorderRadius.circular(Radii.lg),
              border: Border.all(
                color: info.current.color.withValues(alpha: 0.6),
              ),
            ),
            child: Column(
              children: [
                Medal(
                  rank: info.current,
                  size: 140,
                  progress: info.progress,
                  level: level,
                ),
                const SizedBox(height: Spacing.md),
                Text(
                  info.current.name,
                  style: TextStyle(
                    color: info.current.color,
                    fontSize: 24,
                    fontWeight: FontWeight.w900,
                  ),
                ),
                Text(
                  '${t('level')} $level/$maxLevel',
                  style: TextStyle(color: c.textMuted),
                ),
                const SizedBox(height: 6),
                Text(
                  '$pts pt • ${workouts.length} ${t('workouts').toLowerCase()}',
                  style: TextStyle(color: c.text),
                ),
                const SizedBox(height: 6),
                Text(
                  info.next != null
                      ? t.p('points_for_next', {
                          'n': '${info.next!.min - pts}',
                          'name': info.next!.name,
                        })
                      : t('max_rank'),
                  style: TextStyle(color: c.textMuted, fontSize: 12),
                  textAlign: TextAlign.center,
                ),
                const SizedBox(height: Spacing.md),
                OutlinedButton.icon(
                  icon: const Icon(Icons.share),
                  label: Text(t('share_progress')),
                  onPressed: () {
                    FeedbackService.onTap();
                    Share.share(
                      'Sono ${info.current.name} (Lv $level) su CaliStrack con $pts punti! 💪',
                    );
                  },
                ),
              ],
            ),
          ),
          const SizedBox(height: Spacing.lg),
          Text(
            t('ranks_scale'),
            style: TextStyle(color: c.textMuted, fontWeight: FontWeight.w700),
          ),
          const SizedBox(height: Spacing.sm),
          ...ranks.reversed.map((r) {
            final unlocked = pts >= r.min;
            final isCurrent = r.id == info.current.id;
            return Opacity(
              opacity: unlocked ? 1 : 0.55,
              child: Container(
                margin: const EdgeInsets.only(bottom: Spacing.sm),
                padding: const EdgeInsets.all(Spacing.sm),
                decoration: BoxDecoration(
                  color: c.card,
                  borderRadius: BorderRadius.circular(Radii.md),
                  border: Border.all(
                    color: isCurrent ? r.color : c.border,
                    width: isCurrent ? 2 : 1,
                  ),
                ),
                child: Row(
                  children: [
                    Medal(
                      rank: r,
                      size: 54,
                      progress: unlocked ? 1 : 0,
                      locked: !unlocked,
                    ),
                    const SizedBox(width: Spacing.md),
                    Expanded(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text(
                            'Lv ${levelOf(r.id)} · ${r.name}',
                            style: const TextStyle(fontWeight: FontWeight.w700),
                          ),
                          Text(
                            '${r.min} ${t('points')}',
                            style: TextStyle(color: c.textMuted, fontSize: 12),
                          ),
                        ],
                      ),
                    ),
                    if (isCurrent)
                      Container(
                        padding: const EdgeInsets.symmetric(
                          horizontal: 8,
                          vertical: 3,
                        ),
                        decoration: BoxDecoration(
                          color: r.color.withValues(alpha: 0.2),
                          borderRadius: BorderRadius.circular(Radii.sm),
                        ),
                        child: Text(
                          t('current'),
                          style: TextStyle(
                            color: r.color,
                            fontSize: 10,
                            fontWeight: FontWeight.w800,
                          ),
                        ),
                      ),
                  ],
                ),
              ),
            );
          }),
          const SizedBox(height: Spacing.lg),
          Text(
            t('points_explanation'),
            style: TextStyle(color: c.textMuted, fontSize: 12),
          ),
        ],
      ),
    );
  }
}
